src/lib/sticky-header/sticky-header-dir.ts

```diff
@@ -0,0 +1,295 @@
+  public stickyParent: HTMLElement;
+  public upperScrollableContainer: HTMLElement;
+
+  /**
+   * the original css of the sticky element, used to reset the sticky element
+   * when it is being unstuck
+   */
+  public originalCss: any;
+
+  private _containerStart: number;
+  private _scrollFinish: number;
+
+  private _scrollingWidth: number;
+
+  constructor(private _element: ElementRef,
+              public scrollable: Scrollable,
+              @Optional() public parentReg: CdkStickyRegion) {
+    this.element = _element.nativeElement;
+    this.upperScrollableContainer = scrollable.getElementRef().nativeElement;
+    this.scrollableRegion = scrollable.getElementRef().nativeElement;
+    if (parentReg != null) {
+      this.parentRegion = parentReg.getElementRef().nativeElement;
+    }
+  }
+
+  ngAfterViewInit(): void {
+
+    if (this.parentRegion != null) {
+      this.stickyParent = this.parentRegion;
+    }else {
+      this.stickyParent = this.element.parentElement;
+    }
+
+    this.originalCss = {
+      zIndex: this.getCssValue(this.element, 'zIndex'),
+      position: this.getCssValue(this.element, 'position'),
+      top: this.getCssValue(this.element, 'top'),
+      right: this.getCssValue(this.element, 'right'),
+      left: this.getCssValue(this.element, 'left'),
+      bottom: this.getCssValue(this.element, 'bottom'),
+      width: this.getCssValue(this.element, 'width'),
+    };
+
+    this.attach();
+
+    this.defineRestrictionsAndStick();
+  }
+
+  ngOnDestroy(): void {
+    this.upperScrollableContainer.removeEventListener('scroll', this._onScrollBind);
+    this.upperScrollableContainer.removeEventListener('resize', this._onResizeBind);
+    this.upperScrollableContainer.removeEventListener('touchmove', this._onTouchMoveBind);
+  }
+
+  attach() {
+    this.upperScrollableContainer.addEventListener('scroll', this._onScrollBind, false);
+    this.upperScrollableContainer.addEventListener('resize', this._onResizeBind, false);
+
+    // Have to add a 'onTouchMove' listener to make sticky header work on mobile phones
+    this.upperScrollableContainer.addEventListener('touchmove', this._onTouchMoveBind, false);
+  }
+
+  onScroll(): void {
+    this.defineRestrictionsAndStick();
+  }
+
+  onTouchMove(): void {
+    this.defineRestrictionsAndStick();
+  }
+
+  onResize(): void {
+    this.defineRestrictionsAndStick();
+    // If there's already a header being stick when the page is
+    // resized. The CSS style of the sticky-header may be not fit
+    // the resized window. So we need to unstuck it then re-stick it.
+    // unstuck() can set 'isStuck' to FALSE. Then stickElement() can work.
+    if (this.isStuck) {
+      this.unstuckElement();
+      this.stickElement();
+    }
+  }
+
+  /**
+   * define the restrictions of the sticky header(including stickyWidth,
+   * when to start, when to finish)
+   */
+  defineRestrictions(): void {
+    let containerTop: any = this.stickyParent.getBoundingClientRect();
```

Comment:
Yes. It will cause slowdown. But I have to run it on every scroll and resize event. Because I need to use it to get the top of current StickyRegion, and use the value to judge whether the current header need to be stuck or not.
Is there any way to optimize it?